Add fail animation

frenzibyte opened the pull request:
### Implementing a proper fail scene like the stable one instead of a simple popup fail overlay.

[Video Preview](https://drive.google.com/open?id=1G_WH9Q5v6LFzFKaUieq2L4zUYaj2DzN5), Thanks to @Poyo-SSB
- Closes #3361
- Closes #2579
- [x] Depends on ppy/osu-resources#54

frenzibyte committed:
Add proper fail scene

frenzibyte committed:
Make Playfield accessible by DrawableRuleset

Playfield wasn't accessible by DrawableRuleset but DrawableRuleset<T>, here we are fixing this issue

frenzibyte committed:
Unneccesary lines

frenzibyte committed:
Add Gameplay Input Blocker

Used for blocking input while in failing state

frenzibyte committed:
Merge branch 'master' into proper-fail-scene

peppy commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -33,6 +33,11 @@ public class GameplayClockContainer : Container
 
         public readonly BindableBool IsPaused = new BindableBool();
 
+        public override bool PropagateNonPositionalInputSubTree => !BlockGameplayInput;
```
did you try doing this in `RulesetInputManager` as I proposed?

iiSaLMaN replied:
I tried, it disable the cursor movements too...

peppy commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -33,6 +33,11 @@ public class GameplayClockContainer : Container
+        public override bool PropagateNonPositionalInputSubTree => !BlockGameplayInput;
+        public override bool PropagatePositionalInputSubTree => !BlockGameplayInput;
+
+        public bool BlockGameplayInput { get; set; }
```
xmldoc. also there's no way this should be in `GameplayClockContainer` (a container managing a clock, not input).

iiSaLMaN replied:
I see...

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -345,9 +366,42 @@ private bool onFail()
+
+        private void updateFail()
+        {
+            audio.Track.Frequency.Value -= Game.IsActive.Value ? 0.0015 : 0.00075;
```
what is this `Game.IsActive` nonsense? you should be using `Clock.Elapsed` or something similar.

iiSaLMaN replied:
Doesn't the stable version in a fail scene, when the game is not focused, it slows the failing process by half?

peppy replied:
How'd you figure that out?

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -345,9 +366,42 @@ private bool onFail()
+            DrawableRuleset.Playfield.Alpha -= Game.IsActive.Value ? 0.0015f : 0.00075f;
+
+            Random objRand = new Random();
+            foreach (DrawableHitObject Object in DrawableRuleset.Playfield.AllHitObjects.ToList())
```
you are operating on all objects ever. this will not work for long beatmaps.

iiSaLMaN replied:
How about `DrawableRuleset.Playfield.HitObjectContainer.Objects`?

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -345,9 +366,42 @@ private bool onFail()
+
+            DrawableRuleset.Playfield.Alpha -= Game.IsActive.Value ? 0.0015f : 0.00075f;
+
+            Random objRand = new Random();
```
Use `RNG.`. Also this seems wrong, since the random values are going to change every frame?

iiSaLMaN replied:
I'm using RNG now, and i made another calculation that you'll see on the next commit

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -345,9 +366,42 @@ private bool onFail()
+            if (audio.Track.Frequency.Value > 0)
+                return;
+
+            IsFailing = false; // Stop slowing down the audio
```
very unintuitive.

iiSaLMaN replied:
Will be removed

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -345,9 +366,42 @@ private bool onFail()
+            audio.Track.Frequency.Value = 1;
+
+            // Return gameplay input
+            GameplayClockContainer.BlockGameplayInput = false;
```
why?

iiSaLMaN replied:
Will be removed

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -345,9 +366,42 @@ private bool onFail()
+
+            GameplayClockContainer.Stop();
+
+            audio.Track.Frequency.Value = 1;
```
why?

iiSaLMaN replied:
Track's playback speed must be restored for other uses

Aergwyn replied:
What if it wasn't `1` to begin with? Example: playing with Double/Half Time?

iiSaLMaN replied:
Wait, Double Time and Half Time doesn't use Frequency afaik, so it'll be 1 in either way + i'm gonna push a huge change in a few minutes

peppy commented:
If the code remains are intrusive as how you have it, it likely should not be in `Player` itself, but a separate class which manages the transition. The way I was intending this to be done was to use transforms, not manual every-update operations.

iiSaLMaN commented:
Alright, i'll look at it

frenzibyte committed:
Merge branch 'master' into proper-fail-scene

frenzibyte committed:
New design

frenzibyte committed:
Using transform instead of manual.

frenzibyte committed:
Merge branch 'master' into proper-fail-scene

frenzibyte committed:
Merge branch 'proper-fail-scene' of https://github.com/iiSaLMaN/osu into proper-fail-scene

frenzibyte committed:
Fix AppVeyor Errors

frenzibyte committed:
Unnecessary using directive

frenzibyte committed:
Merge branch 'master' into proper-fail-scene

frenzibyte committed:
Unnecessary

frenzibyte committed:
Merge branch 'master' into proper-fail-scene

frenzibyte committed:
Fix exiting from fail scene bug

A bug that causes the song to be stopped if you hold escape or exit while in a fail scene

frenzibyte committed:
Fix bug

frenzibyte committed:
Fix AppVeyor Errors

frenzibyte committed:
Merge branch 'master' into proper-fail-scene

frenzibyte committed:
Merge branch 'master' into proper-fail-scene

peppy commented:
You need to update the existing (failing) tests to match your new behaviour.

peppy committed:
Merge remote-tracking branch 'upstream/master' into proper-fail-scene

peppy committed:
Update resources

iiSaLMaN commented:
I'm just missing one thing other than the test, blocking input..

frenzibyte committed:
Add steps for fail transform

peppy committed:
Merge remote-tracking branch 'Joehuu/master' into proper-fail-scene

peppy commented on osu.Game/Screens/Play/FailTransform.cs:
```diff
@@ -0,0 +1,84 @@
+            clockRate.Default = clockRate.Value;
+            trackFreq.Default = trackFreq.Value;
+
+            sampleFail = audioManager.Sample.Get(@"Gameplay/failsound");
```
This shouldn't be done in the constructor. Consider just moving this part to `Player` (alongside `sampleRestart` which is already there).

peppy commented on osu.Game/Screens/Play/FailTransform.cs:
```diff
@@ -0,0 +1,84 @@
+
+            // TODO: Use RNG with 'seed' if it was implemented.
+            Random randObj = new Random(seed);
+            foreach (DrawableHitObject obj in d.Objects)
```
Should not run on off-screen objects. Will become a performance issue. Use `AliveObjects` instead.

peppy commented on osu.Game/Screens/Play/FailTransform.cs:
```diff
@@ -0,0 +1,84 @@
+            foreach (DrawableHitObject obj in d.Objects)
+            {
+                obj.Rotation += RNG.NextSingle(0.2f);
+                obj.Y += RNG.NextSingle(0.4f, 0.8f);
```
This should be more of a gravity effect, rather than a linear drop. Also I don't think this should be be randomised.

peppy commented on osu.Game/Screens/Play/FailTransform.cs:
```diff
@@ -0,0 +1,84 @@
+            {
+                obj.Rotation += RNG.NextSingle(0.2f);
+                obj.Y += RNG.NextSingle(0.4f, 0.8f);
+                obj.X += (randObj.NextDouble() > 0.5) ? RNG.NextSingle(0.0f, 0.4f) : RNG.NextSingle(-0.4f, -0.0f);
```
In stable I don't believe objects move sideways.

peppy commented on osu.Game/Screens/Play/FailTransform.cs:
```diff
@@ -0,0 +1,84 @@
+            d.Alpha = currentValue;
+
+            // TODO: Use RNG with 'seed' if it was implemented.
+            Random randObj = new Random(seed);
```
I don't get the point of this. You're using `RNG.` inside, which means things are going to change directions and velocities and stuff. Did you mean to use this instead?

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -58,6 +59,9 @@ public class Player : ScreenWithBeatmapBackground
         private ScoreManager scoreManager { get; set; }
 
+        [Resolved]
+        private AudioManager audioManager { get; set; }
```
Hopefully not required with my above proposed changes.

peppy commented on osu.Game/Screens/Play/Player.cs:
```diff
@@ -484,8 +504,16 @@ public override bool OnExiting(IScreen next)
             GameplayClockContainer.ResetLocalAdjustments();
 
+            // Reset track frequency just in case the player left while we are in failing scene
+            // TODO: Resetting Frequency values should be done inside FailTransform when removed since it's the one who modified it, but for some reason the base is not invoking onAbort
```
Worth investigating why this is happening.

peppy committed:
Merge remote-tracking branch 'upstream/master' into proper-fail-scene

peppy commented:
@iiSaLMaN thanks for your contribution here. I used it as a base but ended up rewriting most of the content. My implementation simplifies the transforms, adds tests, fixes a few edge cases and ensures compatibility with all rulesets.

iiSaLMaN commented:
@peppy Pretty cool, i didn't have much time so yeah..
